jupyterlab/commands.py

```diff
@@ -38,6 +38,75 @@
 DEV_DIR = osp.abspath(os.path.join(HERE, '..', 'dev_mode'))
 
 
+class ProgressProcess(Process):
+
+    def __init__(self, cmd, logger=None, cwd=None, kill_event=None,
+                 env=None):
+        """Start a subprocess that can be run asynchronously.
+
+        Parameters
+        ----------
+        cmd: list
+            The command to run.
+        logger: :class:`~logger.Logger`, optional
+            The logger instance.
+        cwd: string, optional
+            The cwd of the process.
+        env: dict, optional
+            The environment for the process.
+        kill_event: :class:`~threading.Event`, optional
+            An event used to kill the process operation.
+        quiet: bool, optional
+            Whether to suppress output.
+        """
+        if not isinstance(cmd, (list, tuple)):
+            raise ValueError('Command must be given as a list')
+
+        if kill_event and kill_event.is_set():
+            raise ValueError('Process aborted')
+
+        self.logger = logger = logger or logging.getLogger('jupyterlab')
+        self._last_line = ''
+        self.cmd = cmd
+        self.logger.debug('> ' + list2cmdline(cmd))
+
+        self.proc = self._create_process(
+            cwd=cwd,
+            env=env,
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            universal_newlines=True,
+        )
+        self._kill_event = kill_event or Event()
+
+        Process._procs.add(self)
+
+    def wait(self):
+        cache = []
+        proc = self.proc
+        kill_event = self._kill_event
+        import itertools
+        spinner = itertools.cycle(['-', '/', '|', '\\'])
+        while proc.poll() is None:
+            sys.stdout.write(next(spinner))   # write the next character
+            sys.stdout.flush()                # flush stdout buffer (actual character display)
+            sys.stdout.write('\b')
+            if kill_event.is_set():
+                self.terminate()
+                raise ValueError('Process was aborted')
+            try:
+                out, _ = proc.communicate(timeout=.1)
+                cache.append(out)
+            except subprocess.TimeoutExpired:
+                continue
+        if proc.returncode:
+            self.logger.error('\n'.join(cache))
```

Comment:
This still prints to the terminal, correct?   Can we instead write this to a local file like the npm-error.log file?   And then have that file get cleared on startup of these apps?

Reply:
In that case, I would rather log it to debug and rely on the file logger described elsewhere. If we're using a log file, it should be as verbose as possible, including the result of previous commands (they might have the key to why this command failed).

Reply:
I'd want it to be cleared after a successful run at least. Otherwise it could get huge an irrelevant.